community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/DataImporter.java

```diff
@@ -0,0 +1,291 @@
+        }
+
+        public long nodesImported()
+        {
+            return this.nodes.sum();
+        }
+
+        public long propertiesImported()
+        {
+            return this.properties.sum();
+        }
+
+        public long relationshipsImported()
+        {
+            return this.relationships.sum();
+        }
+
+        @Override
+        public String toString()
+        {
+            return format( "Imported:%n  %d nodes%n  %d relationships%n  %d properties",
+                    nodes.sum(), relationships.sum(), properties.sum() );
+        }
+    }
+
+    private static long importData( String title, int numRunners, InputIterable data, BatchingNeoStores stores,
+            Supplier<EntityImporter> visitors, ExecutionMonitor executionMonitor, StatsProvider memoryStatsProvider )
+            throws IOException
+    {
+        LongAdder roughEntityCountProgress = new LongAdder();
+        ExecutorService pool = Executors.newFixedThreadPool( numRunners,
+                new NamedThreadFactory( title + "Importer" ) );
+        IoMonitor writeMonitor = new IoMonitor( stores.getIoTracer() );
+        ControllableStep step = new ControllableStep( title, roughEntityCountProgress, Configuration.DEFAULT,
+                writeMonitor, memoryStatsProvider );
+        StageExecution execution = new StageExecution( title, null, Configuration.DEFAULT, Collections.singletonList( step ), 0 );
+        InputIterator dataIterator = data.iterator();
+        for ( int i = 0; i < numRunners; i++ )
+        {
+            pool.submit( new ExhaustingEntityImporterRunnable(
+                    execution, dataIterator, visitors.get(), roughEntityCountProgress ) );
+        }
+        pool.shutdown();
+
+        executionMonitor.start( execution );
+        long startTime = currentTimeMillis();
+        long nextWait = 0;
+        try
+        {
+            while ( !pool.awaitTermination( nextWait, TimeUnit.MILLISECONDS ) )
+            {
+                executionMonitor.check( execution );
+                nextWait = executionMonitor.nextCheckTime() - currentTimeMillis();
+            }
+        }
+        catch ( InterruptedException e )
+        {
+            Thread.currentThread().interrupt();
+            throw new IOException( e );
+        }
+        execution.assertHealthy();
+        step.markAsCompleted();
+        writeMonitor.stop();
+        executionMonitor.end( execution, currentTimeMillis() - startTime );
+        execution.assertHealthy();
+
+        return roughEntityCountProgress.sum();
+    }
+
+    public static void importNodes( int numRunners, Input input, BatchingNeoStores stores, IdMapper idMapper,
+            ExecutionMonitor executionMonitor, Monitor monitor )
+                    throws IOException
+    {
+        importData( NODE_IMPORT_NAME, numRunners, input.nodes(), stores, () ->
+            new NodeImporter( stores, idMapper, monitor ), executionMonitor, new MemoryUsageStatsProvider( stores, idMapper ) );
```

Comment:
How about having methods for creating these lambdas? The NodeImporter one is not so bad, but the RelationshipImporter involves a lot of fields.

Reply:
You mean introduce variable? Sure

Reply:
Variable or method. Either works.